### DISCUSSION TITLE
ModuleNotFoundError when importing python file in main.py

### DISCUSSION THREAD

**Discussion by lennizle** [fralau/mkdocs-macros-plugin#120] [Q&A] [answered]
Hi there, 
I am gratefully using mkdocs macros plugin for documentation my software project! 

My idea was to get Table-definitions (e.g. stored in a list) from another python file in the source code and auto-parse it to html tables. 

Now, when I include an import line at the main.py referencing such file I get a `ModuleNotFoundError: No module named 'utils'`. It works neither with relative nor with absolute paths. It does neither work not in the same nor in different folders. Python processes the line in debug mode. But when I run `$ mkdocs serve` it throws that error. I can provide more information if needed.

Can you please help me solve this?

**Comment by fralau** [comment]
Sure! 

First of all, have you listed the names of the macros you would like to implement? [Did you create a `main.py` file](https://mkdocs-macros-plugin.readthedocs.io/en/latest/#create-your-own-macros-and-filters), as required?

**Reply by lennizle** [reply]
Thank you for answering so quickly @fralau 

Yes my `main.py` file is in the top directory and  set up as ordinary as it can be:  3 `import` lines, then `def define_env(env):` and 1 function inside

So you if say it is generally possible to import files, do you maybe have a working example I can have a look at?

**Comment by fuchsst** [comment]
same problem
i have the following strucure (subfolders in the mkdocs project directory)
- `macros`
  - `__init__.py`
  - `mymacro.py`
  - `util`
    - `helper1.py`
    - `helper2.py`
    - `__init__.py`

the `plugins.macros.module_name` setting in the `mkdocs.yml` is `macros/mymacro`

in the `macros/mymacro.py` the first line imports the first util script
`from util.helper1 import util_func`

but on startup it complains 
```
  File "/home/fuchsst/projects/mkdocs_test/macros/mymacro.py", line 1, in <module>
    from util.helper1 import util_func
ModuleNotFoundError: No module named 'util'
```

to me it looks like it doesn't treat it as a module?

**Comment by fralau** [comment]
I believe I see what's going on. [According to the doc](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/#location-of-the-modules), **what is imported is either a module (file) or a package.** Hence your whole `macros` directory will be considered a **package** by mkdocs-macros; and within a package you must use relative imports:

`from .util.helper1 import util_func`

**Reply by fuchsst** [reply]
when i add the dot for relative import i get `ImportError: attempted relative import with no known parent package`

**Reply by lennizle** [reply]
+1

**Comment by lennizle** [comment]
Success to report!!! So if `mkdocs.yml` setting is a folder e.g.  `module_name: utils` ,  the `define_env()` function is accessible through the `__init__.py` file. Now, in the init, imports from within the package `utils` need to be relative, outside of the folder absolute paths. It works 💯 . Thanks to all!

**Reply by fuchsst** [reply]
mh, when i refer to the folder and put the `def define_env(env)` in the `__init__.py` file i get
`ModuleNotFoundError: No module named 'macros'`

can you post a full example, so maybe i see then what i do wrong?

**Reply by lennizle** [reply]
mhh maybe put `macros` outside of `docs` folder. make sure next to `__init__.py` there is another file in the folder. This made it work in my case

**Reply by fuchsst** [reply]
it's parallel to the docs folder :thinking:

**Comment by fralau** [comment, marked as answer]
What it tells me, is that I should put more information on file structure, on the [modules page](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/). 🤔

**Reply by fuchsst** [reply]
yeah, if you have a complete example, that would be great...don't get where i miss something
my plan is to have a couple of scripts in sub-folder of a `macros` folder that is located in the mkdocs project root. in there i have a main.py with the `def define_env(env)`...each call there is basically just a call of another `def` that resides in a python script in subfolders in the macros directory
but for some reason it doesn't find them. for now i can put just everything in one file, but i know it'll grow and become to big....

**Reply by fralau** [reply]
There are [functioning examples provided with the source code](https://github.com/fralau/mkdocs_macros_plugin/tree/master/test), perhaps you could find indications in them?

**Reply by fuchsst** [reply]
not really, because none of the example python scripts has an import of another python file that is part of the project (only python standard libs like `os`)

**Reply by fralau** [reply]
@fuchsst Would you be able to offer a complete, minimal example of what you are trying to do? In that way, I could download and test it.

**Reply by lennizle** [reply]
@fuchsst how I managed to do it is as follows: (@fralau pls correct me if there is a smoother way)
```
macros
├── __init__.py
├── utils.py
├── model.py
└── logic.py
```

in __init__.py is the **only** `def define_env(env)` function across the repo. Within the @env.macros function you call functions that you import from other py files:

```
from .model import list_of_tables,meta_header,data_header
from .logic import create_zip_file
from .model import create_md_table

def define_env(env):
    @env.macro
    def download_as_xlsx():
        s = create_zip_file()
        return s

    @env.macro
    def create_tables():
        s = create_md_table(list_of_tables,meta_header,data_header,'pipe')
        return s

    ....
```

**Comment by fralau** [comment]
@lennizle That seems formally perfect to me, exactly as intended! If you agree, I would like to put your example in the documentation.
